src/bin/pgcopydb/env_utils.c

```diff
@@ -215,3 +219,132 @@ get_env_pgdata_or_exit(char *pgdata)
 			  "or from --pgdata");
 	exit(EXIT_CODE_BAD_ARGS);
 }
+
+
+/*
+ * get_env_using_parsers iterates over the parsers array and calls
+ * get_env_using_parser for each parser.
+ */
+bool
+get_env_using_parsers(EnvParserArray *parsers)
+{
+	int errors = 0;
+
+	for (int i = 0; i < parsers->count; i++)
+	{
+		if (!get_env_using_parser(&parsers->array[i]))
+		{
+			++errors;
+		}
+	}
+
+	return errors == 0;
+}
+
+
+/*
+ * get_env_using_parser checks if the environment variable exists and if it does
+ * it parses it and stores it in the target. If the environment variable is not
+ * set, the target is not modified. If the environment variable is set but the
+ * parsing fails, an error is logged and the errors counter is incremented.
+ */
+bool
+get_env_using_parser(EnvParser *parser)
+{
+	if (!env_exists(parser->envname))
+	{
+		return true;
+	}
+
+	switch (parser->type)
+	{
+		case ENV_TYPE_INT:
+		{
+			char envValue[BUFSIZE] = { 0 };
+			if (!get_env_copy(parser->envname, envValue, sizeof(envValue)))
+			{
+				/* errors have already been logged */
+				return false;
+			}
+			else if (!stringToInt(envValue, (int *) parser->target) ||
+					 (parser->lowerBounded &&
+					  *((int *) parser->target) < parser->minValue) ||
+					 (parser->upperBounded &&
+					  *((int *) parser->target) > parser->maxValue))
+			{
+				PQExpBuffer errorMessage = createPQExpBuffer();
+
+				appendPQExpBuffer(errorMessage,
+								  "Failed to parse \"%s\": \"%s\", "
+								  "expected an integer",
+								  parser->envname,
+								  envValue);
+
+				if (parser->lowerBounded)
+				{
+					appendPQExpBuffer(errorMessage,
+									  " greater than %d",
```

Comment:
or equal?

Reply:
Would you prefer

`greater than or equal to << parser->minValue >>`

or

`greater than << parser->minValue - 1 >>` ?

In my opinion, it is better to keep the number of characters small. Hence I added `+/- 1` instead of adding `or equal` phrases.

Reply:
Let me share an example. I created 2 sample error messages. I like the first one more, but the other one is also ok.

```
Failed to parse "PGCOPYDB_RESTORE_JOBS": "1234", expected an integer greater than 0 and smaller than 129
```

```
Failed to parse "PGCOPYDB_RESTORE_JOBS": "1234", expected an integer greater than or equal to 1 and smaller than or equal to 128
```

Reply:
What about using math operators `<=` and `>=` ?

Reply:
`Failed to parse "PGCOPYDB_RESTORE_JOBS": "1234", expected an integer >= 1 and <= 128`

It is definitely shorter now. I do not prefer having a binary operator mixed in a natural language. Especially when the left side is missing an expression.

Nevertheless, it is no big deal as it conveys the message clearly either way.

Reply:
Another idea:

`Failed to parse "PGCOPYDB_RESTORE_JOBS": "1234", expected an integer i that satisfies 1 <= i <= 128`

That feels even weirder. Never mind.

PostgreSQL usually shows the min and max values in the detail or hint of the error message. I don't think  this is an option for us. I'd rather have the information on the same line as our users fail to read the consecutive lines of log messages quite frequently.

```c
 ereport(ERROR,
		(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
			errmsg("value %s out of bounds for option \"%s\"",
				value, option->gen->name),
			errdetail("Valid values are between \"%d\" and \"%d\".",
					optint->min, optint->max)));
```

Reply:
Mixing with operators helps quick-reading through the logs, even if it somewhat looks weird. I think I like the `1 <= jobs <= 128` best for the quick-reading aspect to it, even though it's the worst option in terms of being a “proper” English phrase.